warp-terminal: added passthru update script

imadnyc opened the pull request:
## Description of changes

<!--
For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->
Finally got around to adding a passthru update script. Here, I manually updated the macOS hash because to my understanding `update-source-version` expects a platform+architecture, but doesn't accept a general platform, which warp does with it's `.dmg` (again, to my understanding, I don't have a mac). I also wanted to add myself as a maintainer of this, but I already have that commit in #291443, and didn't want to duplicate the commit. If this is ready to be merged before that, though, I'll remove that one and add it here. 
## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) (or backporting [23.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md) and [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

imadnyc committed:
warp-terminal: added passthru update script

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/does-update-source-version-have-a-generic-system-flag/40679/1

a-n-n-a-l-e-e commented on pkgs/by-name/wa/warp-terminal/update.sh:
```diff
@@ -0,0 +1,30 @@
+macOS_hash=$(nix-prefetch-url --type sha256 $macOS_url)
+
+# this monstrosity is needed to find the second appearance of [ hash = "..."; ] and replace it with the correct hash
+sed -i '/hash = ".*";/{G;s/\nX\{1\}//;tend;x;s/^/X/;x;P;d};p;d;:end;s/hash = ".*"/hash = "'"$macOS_hash"'"/;:a;n;ba' "$dirname/package.nix"
```
you can put the hash and version in a json file and have your update script read / write the json file, and have the derivation import the json.

see:
https://github.com/NixOS/nixpkgs/blob/268690ff59e25a06a7c782900144df92ed545584/pkgs/servers/invidious/default.nix


```json
{
	"darwin" : {
		"hash" : "sha256-tFtoD8URMFfJ3HRkyKStuDStFkoRIV97y9kV4pbDPro=",
		"version" : "0.2024.02.20.08.01.stable_01"
	},
	"linux" : {
		"hash" : "sha256-L8alnqSE4crrDozRfPaAAMkLc+5+8d9XBKd5ddsxmD0=",
		"version" : "0.2024.02.20.08.01.stable_01"
	}
}
```

then in nix
```nix
let
pname = "warp-terminal";
versions = lib.importJSON ./versions.json;

linux = stdenv.mkDerivation (finalAttrs:  {
  inherit pname meta;
  inherit (versions.linux) version;
  src = fetchurl {
    inherit (versions.linux) hash;
    url = "https://releases.warp.dev/stable/v${finalAttrs.version}/warp-terminal-v${finalAttrs.version}-1-x86_64.pkg.tar.zst";
  };
[...]
```

imadnyc replied:
Thanks for the writeup! I really appreciate the explanation instead of just changing it. I have the fixed up version that should be ready.

a-n-n-a-l-e-e reviewed (commented):
change looks good. minor comments inline.
do you want to add yourself to the maintainers?

imadnyc commented on pkgs/by-name/wa/warp-terminal/update.sh:
```diff
@@ -0,0 +1,37 @@
+    jq --arg x "$2" "$1 = \$x" < "$dirname/versions.json" | sponge "$dirname/versions.json"
+}
+
+# it's fine to use just the linux URL because the package also assumes the versioning is the same between darwin and linux
+version=$(curl -Ls -o /dev/null -w %{url_effective} https://app.warp.dev/download\?package\=pacman 2>&1 | grep -oP -m 1 "(?<=v)[\d.\w]+" | head -1)
```
I decided to keep it in because it's writing the same version to both, though if it's confusing I could remove it. And I tried not pulling the entire file, but otherwise it wouldn't get the final destination url. the `-I` flag is what that's supposed to do but it doesn't follow through all the redirects.

a-n-n-a-l-e-e replied:
> I decided to keep it in because it's writing the same version to both, though if it's confusing I could remove it. And I tried not pulling the entire file, but otherwise it wouldn't get the final destination url. the `-I` flag is what that's supposed to do but it doesn't follow through all the redirects.

there is only one redirect. not all servers support HEAD, which is what -I is doing, and this one doesn't.

having multiple variables containing the same value is sloppy. granted in this case it is pretty minor.

a-n-n-a-l-e-e replied:
here is a function that will resolve the url. started with `http` rather than `https` to get 2 redirects.

```bash
#!/usr/bin/env bash

set -e

err() {
    echo "$*" >&2
}

resolve_url() {
    local url="$1"
    local -i i max_retries
    ((max_retries = 30))
    url="$1"
    for ((i = 0; i < max_retries; i++)); do
        echo "round $i"
        url="$(curl -s -o /dev/null -w %{redirect_url} "${url}")"
        [[ "$url" != *stable/v* ]] || break
    done
    ((i < max_retries)) || { err "too many retries"; exit 1; }
    echo $url
}

resolve_url "http://app.warp.dev/download?package=pacman"
```

imadnyc replied:
Wow, thanks! I didn't even know about servers not supporting HEAD. And yeah, I don't know why I didn't wrap the JSON stuff in a loop.

imadnyc commented:
Yeah, I'd like to be a maintainer if my other PRS aren't going to be merged before this. I'll update the other ones if this gets merged.

a-n-n-a-l-e-e commented:
> Yeah, I'd like to be a maintainer if my other PRS aren't going to be merged before this. I'll update the other ones if this gets merged.

add a commit to this PR and do it so this PR can get merged.

you need to fix your other commit message though. the body is not adding anything useful to someone who will be looking at the commit logs sometime in the future.

look at the output from `git log` for some examples.

imadnyc committed:
maintainers: add imadnyc

imadnyc commented:
I've added myself to the maintainer list and removed my commit body - I was squashing instead of amending and forgot to remove the commit messages getting added to the body.

a-n-n-a-l-e-e commented:
there are some somewhat strict [rules](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#commit-conventions)  about commit messages. the commit to the maintainers needs to be:
    `maintainers: add imadnyc`
    `warp-terminal: added passthru update script`

the maintainers needs to come first, otherwise if someone is performing a bisect if the eval  will break if they land on the `warp-terminal` change because you are missing from the maintainers. you'll need to do something like `git rebase -i HEAD~2` reorder the commit and rewrite the maintainer message.

imadnyc commented:
I'm gonna pretend it was because I didn't drink coffee today and nothing else. I think it should be good for real this time

a-n-n-a-l-e-e reviewed (approved):
LGTM. validated `curl -s https://api.github.com/users/imadnyc|jq '{name,github:.login,githubId:.id}'` matches github info in maintainers

a-n-n-a-l-e-e commented on pkgs/by-name/wa/warp-terminal/update.sh:
```diff
@@ -0,0 +1,60 @@
+    version=$(get_version "${url}")
+    if [[ ${version} != "$(json_get ".${sys}.version")" ]];
+        then
+            sri=$(nix hash to-sri --type sha256 "$(nix-prefetch-url --type sha256 "${url}")")
```
haven't seen an update posted yet. this might need to be `nix-hash --to-sri` as `nix hash` is experimental and can generate an error if the magic config settings are not set.

a-n-n-a-l-e-e replied:
or not -- i see `nix hash` in some other update.sh. `pkgs/tools/video/recyclarr/update.sh` and looks like it had an auto-update. https://github.com/NixOS/nixpkgs/pull/267039
